ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearch.ts

```diff
@@ -0,0 +1,140 @@
+				return 'search-marker';
+			} else if (match) {
+				stream.pos = match.index;
+			} else {
+				stream.skipToEnd();
+			}
+			return null;
+		} };
+	}
+
+	// Highlights the currently active found work
+	// It's possible to get tricky with this fucntions and just use findNext/findPrev
+	// but this is fast enough and works more naturally with the current search logic
+	function highlightSearch(cm: any, searchTerm: RegExp, index: number) {
+		const marks: any = [];
+
+		const cursor = cm.getSearchCursor(searchTerm);
+
+		let match = null;
+		for (let j = 0; j < index + 1; j++) {
+			if (!cursor.findNext()) {
+				// If we run out of matches then just highlight the final match
+				break;
+			}
+			match = cursor.pos;
+		}
+
+		if (match) {
+			marks.push(cm.markText(match.from, match.to, { className: 'cm-search-marker-selected' }));
+			cm.scrollIntoView(match);
+		}
+
+		return marks;
+	}
+
+	function getSearchTerm(keyword: any) {
+		let searchTerm = new RegExp(keyword.value, 'g');
+
+		if (keyword.accuracy === 'partially') {
+			searchTerm = new RegExp(keyword.value, 'gi');
+		}
+
+		return searchTerm;
+	}
+
+	CodeMirror.defineExtension('setMarkers', function(keywords: any, options: any) {
+		if (!options) {
+			options = { selectedIndex: 0 };
+		}
+
+		clearMarkers();
+
+		// HIGHLIGHT KEYWORDS
+		// When doing a global search it's possible to have multiple keywords
+		// This means we need to highlight each one
+		let marks: any = [];
+		for (let i = 0; i < keywords.length; i++) {
+			const keyword = keywords[i];
+
+			if (keyword.value === '') continue;
+
+			const searchTerm = getSearchTerm(keyword);
+
+			marks = marks.concat(highlightSearch(this, searchTerm, options.selectedIndex));
+		}
+
+		setMarkers(marks);
+
+		// SEARCHOVERLAY
+		// We only want to highlight all matches when there is only 1 search term
+		if (keywords.length !== 1 || keywords[0].value == '') {
+			clearOverlay(this);
+			return 0;
+		}
+
+		const searchTerm = getSearchTerm(keywords[0]);
+
+		// Determine the number of matches in the source, this is passed on
+		// to the NoteEditor component
+		const regexMatches = this.getValue().match(searchTerm);
+		const nMatches = regexMatches ? regexMatches.length : 0;
+
+		// Don't bother clearing and re-calculating the overlay if the search term
+		// hasn't changed
+		if (keywords[0].value === previousKeywordValue) return nMatches;
+
+		clearOverlay(this);
+		setPreviousKeywordValue(keywords[0].value);
+
+		if (overlayTimeout) {
+			clearTimeout(overlayTimeout);
+		}
+		// These operations are pretty slow, so we won't add use them until the user
+		// has finished typing, 500ms is probably enough time
+		const timeout = setTimeout(() => {
+			const scrollMarks = this.showMatchesOnScrollbar(searchTerm, true, 'cm-search-marker-scrollbar');
+			const overlay = searchOverlay(searchTerm);
+			this.addOverlay(overlay);
+			setOverlay(overlay);
+			setScrollbarMarks(scrollMarks);
+		}, 500);
+
+		setOverlayTimeout(timeout);
```

Comment:
I wonder what happens if the editor or extension is gone by the time this timeout runs? Is there some variable you can check inside the timeout like `this.isValid()` to know if the `this` reference is still valid?

Alternative, doesn't CodeMirror provides a way to run clean up code when an extension is unmounted (a bit like effects do)? In which case you could clear the timeout at that point.

Reply:
Thats a good point, I did some testing and the `this` reference is still valid but because it's been unmounted by react this code causes an error.

Unfortunately codemirror doesn't provide support for this sort of clean up. Unrelated I was planning on adding a bit of a framework around the codemirror plugins because currently they duplicate some behavior (such as multi-line actions). Additionally Joplin supports 2 modes (html and markdown) which currently isn't handled by the existing plugins. I plan to make some improvements in that regard (I've mentioned this before but haven't had any time to work on it). I'd like to make some progress there and I'll take into consideration that we need a sort of cleanup code.

Reply:
I guess the useEditorSearch effect is going to be unmounted when CodeMirror gets unmounted too? In that case, for now a simple `useEffect(() => { clearTimeout(overlayTimeout) }, [])` would be enough? Maybe also add a try/catch block in the setTimeout block for extra safety and I guess that should be enough.

> Additionally Joplin supports 2 modes (html and markdown) which currently isn't handled by the existing plugins.

I've checked that and it seems to handle it fine? At least for me it loads the HTML code as expected and highlights it. Is there something more you wanted to support?

Reply:
The codemirror instance isn't handled by react, so `useEffect` doesn't trigger on an unmount. In this instance we should probably manually call some cleanup code when codemirror is unmounted.

> I've checked that and it seems to handle it fine? At least for me it loads the HTML code as expected and highlights it. Is there something more you wanted to support?
I was thinking that some of the utils (like the markdown list handling) shouldn't be used on html notes, but I guess there isn't any reason they would cause issues. Probably better not to over engineer it (if it does cause issues we can just put a guard in the specific utils file).
I think there is probably still room to add a cleanup system and some multicursor helper functions.

Reply:
I see that you've already added some code to create, then clean up the CodeMirror instance here: https://github.com/laurent22/joplin/blob/3807317e6678190eae59496844abb7231e51379d/ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/Editor.tsx#L199 So even though it's not directly handled by React it's controlled by your code, which is what we'd want anyway.

Effects are unmounted in the order they were added, so it means that when Editor gets unmounted, first it would run the clean up code of `useEditorSearch`, then the clean up code to remove the Code Mirror instance, and I think with that we're covered.

> The codemirror instance isn't handled by react, so useEffect doesn't trigger on an unmount.

So I'm not sure what you meant by that? For sure the useEffect clean up code is ran when a component is unmounted.

Reply:
I did a quick test where I added `useEffect` to useEditorSearch, but the code inside the useEffect block was never called. I'll double check as it's possible that I forgot to run watch.